pkg/certificates/certificates_test.go

```diff
@@ -0,0 +1,290 @@
+			Token: [tokenLength]byte{1, 2, 3},
+		})
+	}
+	authsBytes, err := expectedAuths.Marshal()
+	if !assert.NoError(t, err) {
+		t.Fatal(err)
+	}
+	err = authDB.DB.Put(storage.Key("user@example.com"), authsBytes)
+	if !assert.NoError(t, err) {
+		t.Fatal(err)
+	}
+
+	cases := []struct {
+		testID,
+		email string
+		result Authorizations
+	}{
+		{
+			"Non-existant email",
+			"nouser@example.com",
+			nil,
+		},
+		{
+			"Exising email",
+			"user@example.com",
+			expectedAuths,
+		},
+	}
+
+	for _, c := range cases {
+		t.Run(c.testID, func(t *testing.T) {
+			auths, err := authDB.Get(c.email)
+			assert.NoError(t, err)
+			if c.result != nil {
+				assert.NotEmpty(t, auths)
+				assert.Len(t, auths, len(c.result))
+			} else {
+				assert.Empty(t, auths)
+			}
+		})
+	}
+}
+
+func TestNewAuthorization(t *testing.T) {
+	auth, err := NewAuthorization()
+	assert.NoError(t, err)
+	assert.NotNil(t, auth)
+	assert.NotZero(t, auth.Token)
+}
+
+func TestAuthorizations_Marshal(t *testing.T) {
+	t1 := [tokenLength]byte{1, 2, 3}
+	t2 := [tokenLength]byte{4, 5, 6}
+	expectedAuths := Authorizations{
+		{Token: t1},
+		{Token: t2},
+	}
+
+	authsBytes, err := expectedAuths.Marshal()
+	assert.NoError(t, err)
+	assert.NotEmpty(t, authsBytes)
+
+	var actualAuths Authorizations
+	decoder := gob.NewDecoder(bytes.NewBuffer(authsBytes))
+	err = decoder.Decode(&actualAuths)
+	assert.NoError(t, err)
+	assert.NotNil(t, actualAuths)
+	assert.Equal(t, expectedAuths, actualAuths)
+}
+
+func TestAuthorizations_Unmarshal(t *testing.T) {
+	t1 := [tokenLength]byte{1, 2, 3}
+	t2 := [tokenLength]byte{4, 5, 6}
+	expectedAuths := Authorizations{
+		{Token: t1},
+		{Token: t2},
+	}
+
+	authsBytes, err := expectedAuths.Marshal()
+	assert.NoError(t, err)
+	assert.NotEmpty(t, authsBytes)
+
+	var actualAuths Authorizations
+	err = actualAuths.Unmarshal(authsBytes)
+	assert.NoError(t, err)
+	assert.NotNil(t, actualAuths)
+	assert.Equal(t, expectedAuths, actualAuths)
+}
+
+func TestAuthorizations_Group(t *testing.T) {
+	t1 := [tokenLength]byte{1, 2, 3}
+	t2 := [tokenLength]byte{4, 5, 6}
+	auths := make(Authorizations, 10)
+	for i := 0; i < 10; i++ {
+		if i%2 == 0 {
+			auths[i] = &Authorization{
+				Token: t1,
+				Claim: &Claim{
+					Timestamp: time.Now().Unix(),
+				},
+			}
+		} else {
+			auths[i] = &Authorization{
+				Token: t2,
+			}
+		}
+	}
+
+	claimed, open := auths.Group()
+	for _, a := range claimed {
+		assert.NotNil(t, a.Claim)
+	}
+	for _, a := range open {
+		assert.Nil(t, a.Claim)
+	}
+}
+
+func TestAuthorizationDB_Emails(t *testing.T) {
+	ctx := testcontext.New(t)
+	authDB, err := newTestAuthDB(ctx)
+	if !assert.NoError(t, err) {
+		t.Fatal(err)
+	}
+	defer func() {
+		_ = authDB.Close()
+		ctx.Cleanup()
+	}()
+
+	var authErrs utils.ErrorGroup
+	for i := 0; i < 5; i++ {
+		_, err := authDB.Create("user%s@example.com", 1)
+		if err != nil {
+			authErrs.Add(err)
+		}
+	}
+	if err = authErrs.Finish(); !assert.NoError(t, err) {
```

Comment:
Why `!assert.NoError(t, err)` instead of `assert.Error(t, err)`?

Reply:
I think you should split this into two separate lines instead of doing multiple different things in the if statement

Reply:
Agreed to split that line. To answer your question: I'm simultaneously making an assertion that there is no error while adding behavior in the event that there is. If I were to do `assert.Error(t, err)` instead, the assertion would always fail if the test were otherwise passing.
```
=== RUN   TestAuthorizationDB_Emails
--- FAIL: TestAuthorizationDB_Emails (0.15s)
    certificates_test.go:260: 
        	Error Trace:	certificates_test.go:260
        	Error:      	An error is expected but got nil.
        	Test:       	TestAuthorizationDB_Emails
FAIL
```